### PR TITLE
Using MSI and Role Assignments to manage Managed Application resources

### PR DESCRIPTION
**Pull request by raosuhas**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [ ] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        "label": "Application settings",
+        "subLabel": {
+          "preValidation": "Configure the application settings",
+          "postValidation": "Done"
+        },
+        "bladeTitle": "Application settings",
+        "elements": [
+          {
+            "name": "storageAccount",
+            "type": "Microsoft.Common.TextBox",
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json]: use the storageAccount control for this - be sure to set hideExisting to true and output all the properties of the control

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            },
+            "visible": true
+          }
+        ]
+      }
+    ],
+    "outputs": {
+      "storageAccountName": "[steps('applicationSettings').storageAccount.name]",
+      "logicAppName": "[steps('applicationSettings').storageAccount.name]",
+      "rolenameguid": "[guid()]",
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json]: remove this and move guid generation to the template

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        "logicAppName": {
+            "type": "string",
+            "defaultValue": "logicappwf"
+        },
+        "rolenameguid": {
+            "type": "string"
+        },
+        "storageAccountName": {
+            "type": "string",
+            "defaultValue": "suhas"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: remove defaultValue

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        }
+    },
+    "variables": {
+        "customResourceProviderName": "public",
+        "storageAccountType": "Standard_LRS",
+        "_deploymentLink": "[deployment().properties.templateLink.uri]",
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: this is never used

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+        "storageBlobOwnerId": "[concat('/subscriptions/', subscription().subscriptionId, '/providers/Microsoft.Authorization/roleDefinitions/', 'b7e6dc6d-f1e8-4753-8033-0f276bb0955b')]"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: remove concat and use subscriptionResourceId() to create the resourceId

File: 201-managed-application-usingmsi-for-management/README.md [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+<a href="https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F201-managed-application-usingmsi-for-management%2Fazuredeploy.json" target="_blank">
+    <img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/deploytoazure.png"/>
+</a>
+<a href="http://armviz.io/#/?load=https%3A%2F%2Fraw.githubusercontent.com%2FAzure%2Fazure-quickstart-templates%2Fmaster%2F201-managed-application-usingmsi-for-management%2Fazuredeploy.json" target="_blank">
+    <img src="https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/1-CONTRIBUTION-GUIDE/images/visualizebutton.png"/>
+</a>
+
+This sample demonstrates how you can include MSI and role assignments in your managed application template to manage and modify resources within the managed resource group. It also includes a sample action on the managed application which can be used to trigger these actions by the user of the managed application. To understand this sample we assume you are aware of the following concepts (Links with samples)
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/README.md]: add badges (see CI results for details)

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+            },
+            "properties": {
+                "state": "Enabled",
+                "definition": {
+                    "$schema": "https://schema.management.azure.com/providers/Microsoft.Logic/schemas/2016-06-01/workflowdefinition.json#",
+                    "actions": {
+                        "HTTP_Request_To_create_Container_in_provided_storage_account": {
+                            "inputs": {
+                                "authentication": {
+                                    "audience": "https://@{coalesce(triggerBody()?['properties']?['storageAccountName'], parameters('storagename'))}.blob.core.windows.net",
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: use reference() to get the storage endpoint

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                                    "type": "ManagedServiceIdentity"
+                                },
+                                "headers": {
+                                    "x-ms-version": "2017-11-09"
+                                },
+                                "method": "PUT",
+                                "uri": "https://@{coalesce(triggerBody()?['properties']?['storageAccountName'],parameters('storagename'))}.blob.core.windows.net/@{triggerBody()?['properties']?['containerName']}?restype=container"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: use reference() to get the uri

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+            "apiVersion": "2018-09-01-preview",
+            "name": "[variables('customResourceProviderName')]",
+            "location": "[parameters('location')]",
+            "properties": {
+                "actions": [
+                    {
+                        "name": "createContainer",
+                        "mode": "Secure",
+                        "routingType": "Proxy",
+                        "endpoint": "[listCallbackURL(concat(resourceId('Microsoft.Logic/workflows', parameters('logicAppName')), '/triggers/CustomProviderWebhook'), '2017-07-01').value]"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: remove concat() and use use resourceId()

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+        {
+            "apiVersion": "2018-07-01",
+            "type": "Microsoft.Authorization/roleAssignments",
+            "name": "[parameters('rolenameguid')]",
+            "properties": {
+                "roleDefinitionId": "[variables('storageBlobOwnerId')]",
+                "principalId": "[reference(parameters('logicAppName'), '2017-07-01', 'Full').identity.principalId]",
+                "scope": "[resourceGroup().id]"
+            },
+             "dependsOn": [
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: move this above properties

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+                "actions": [
+                    {
+                        "name": "createContainer",
+                        "mode": "Secure",
+                        "routingType": "Proxy",
+                        "endpoint": "[listCallbackURL(concat(resourceId('Microsoft.Logic/workflows', parameters('logicAppName')), '/triggers/CustomProviderWebhook'), '2017-07-01').value]"
+                    }
+                ]
+            },
+            "dependsOn": [
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: move this above properties

File: 201-managed-application-usingmsi-for-management/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,97 @@
+{
+    "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "location": {
+            "type": "string",
+            "allowedValues": [ "australiaeast", "eastus", "westeurope" ],
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/azuredeploy.json]: remove allowedValues

File: 201-managed-application-usingmsi-for-management/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+        "logicAppName": {
+            "type": "string",
+            "defaultValue": "logicappwfsdc"
+        },
+        "rolenameguid": {
+            "type": "string"
+        },
+        "storageAccountName": {
+            "type": "string",
+            "defaultValue": "containerstor"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/azuredeploy.json]: remove defaultValue

File: 201-managed-application-usingmsi-for-management/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+            "defaultValue": ""
+        }
+    },
+    "variables": {
+        "lockLevel": "None",
+        "description": "Sample Managed application definition",
+        "displayName": "Sample Managed application definition",
+        "managedApplicationDefinitionName": "[concat(parameters('applicationName'), '_ApplicationDefinition')]",
+        "packageFileUri": "[uri(parameters('_artifactsLocation'), concat('artifacts/ManagedAppZip/pkg.zip', parameters('_artifactsLocationSasToken')))]",
+        "managedResourceGroupId": "[concat(subscription().id,'/resourceGroups/', concat(parameters('applicationName'),'_managed'))]",
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/azuredeploy.json]: remove concat() and use resourceId()

File: 201-managed-application-usingmsi-for-management/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+            }
+        },
+        {
+            "type": "Microsoft.Solutions/applications",
+            "apiVersion": "2019-07-01",
+            "name": "[parameters('applicationName')]",
+            "location": "[parameters('location')]",
+            "kind": "ServiceCatalog",
+            "dependsOn": [
+                "[concat('Microsoft.Solutions/applicationDefinitions/',variables('managedApplicationDefinitionName'))]"
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/azuredeploy.json]: use resourceId()

File: 201-managed-application-usingmsi-for-management/metadata.json [hunk 16]
```diff
@@ -0,0 +1,14 @@
+{
```
Comment (bmoore-msft) [inline 201-managed-application-usingmsi-for-management/metadata.json]: validate this file against the schema and fix errors (see CI results)

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json (hunk 1 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json]: Fixed

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json (hunk 2 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/createUiDefinition.json]: Fixed

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json (hunk 4 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: Removed

File: 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json (hunk 5 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/artifacts/ManagedAppZip/mainTemplate.json]: Done

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 12 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: Custom providers is only available in some regions I have to restrict this value, how can i do that ? 
we already do this for other custom provider samples. Check those here : 
https://github.com/Azure/azure-quickstart-templates/blob/master/101-custom-rp-with-function/azuredeploy.json

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 13 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: Done

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 14 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: This is a resource group id. I could not find how to use a function to create this in a template

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 15 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: used the variable created above.

File: 201-managed-application-usingmsi-for-management/metadata.json (hunk 16 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/metadata.json]: Fixed

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 14 above)
Comment (bmoore-msft) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: use subscriptionResourceId()

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 12 above)
Comment (bmoore-msft) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: ok, but please keep an eye on this, its a very small set of regions and I suspect will become stale quickly

File: 201-managed-application-usingmsi-for-management/azuredeploy.json (hunk 14 above)
Comment (raosuhas) [reply 201-managed-application-usingmsi-for-management/azuredeploy.json]: That does not work either cause that function needs resource type. I tried doing :
"managedResourceGroupId": "[subscriptionResourceId('resourceGroups', concat(parameters('applicationName'),'_managed'))]",

But that fails with error :

Unable to evaluate template language function 'subscriptionResourceId': function requires exactly one multi-segmented argument which must be resource type including resource
provider namespace

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**